discovery/eureka/eureka.go

```diff
@@ -0,0 +1,221 @@
+	appInstanceDataCenterInfoMetadataAvailabilityZone = appInstanceDataCenterInfoMetadataPrefix + "availability_zone"
+	appInstanceDataCenterInfoMetadataInstanceID       = appInstanceDataCenterInfoMetadataPrefix + "instance_id"
+	appInstanceDataCenterInfoMetadataPublicIpv4       = appInstanceDataCenterInfoMetadataPrefix + "public_ipv4"
+	appInstanceDataCenterInfoMetadataPublicHostname   = appInstanceDataCenterInfoMetadataPrefix + "public_hostname"
+	appInstanceDataCenterInfoMetadataAmiManifestPath  = appInstanceDataCenterInfoMetadataPrefix + "ami_manifest_path"
+	appInstanceDataCenterInfoMetadataLocalIpv4        = appInstanceDataCenterInfoMetadataPrefix + "local_ipv4"
+	appInstanceDataCenterInfoMetadataHostname         = appInstanceDataCenterInfoMetadataPrefix + "hostname"
+	appInstanceDataCenterInfoMetadataAmiID            = appInstanceDataCenterInfoMetadataPrefix + "ami_id"
+	appInstanceDataCenterInfoMetadataInstanceType     = appInstanceDataCenterInfoMetadataPrefix + "instance_type"
+	appInstanceCountryIDLabel                         = metaAppInstanceLabelPrefix + "country_id"
+	appInstanceIDLabel                                = metaAppInstanceLabelPrefix + "id"
+	appInstanceMetadataPrefix                         = metaAppInstanceLabelPrefix + "metadata_"
+)
+
+// DefaultSDConfig is the default Eureka SD configuration.
+var DefaultSDConfig = SDConfig{
+	RefreshInterval: model.Duration(30 * time.Second),
+}
+
+// SDConfig is the configuration for applications running on Eureka.
+type SDConfig struct {
+	Server           string                       `yaml:"server,omitempty"`
+	RefreshInterval  model.Duration               `yaml:"refresh_interval,omitempty"`
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if len(c.Server) == 0 {
+		return errors.New("eureka_sd: empty or null eureka server")
+	}
+	_, err = url.Parse(c.Server)
+	if err != nil {
+		return err
+	}
+
+	return c.HTTPClientConfig.Validate()
+}
+
+type applicationsClient func(ctx context.Context, server string, client *http.Client) (*Applications, error)
+
+// Discovery provides service discovery based on a Eureka instance.
+type Discovery struct {
+	*refresh.Discovery
+	client     *http.Client
+	server     string
+	appsClient applicationsClient
+}
+
+// New creates a new Eureka discovery for the given role.
+func NewDiscovery(conf SDConfig, logger log.Logger) (*Discovery, error) {
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "eureka_sd", false)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		client:     &http.Client{Transport: rt},
+		server:     conf.Server,
+		appsClient: fetchApps,
```

Comment:
why do we need this?

Reply:
It was created to separate Eureka's app list inquiry.

Reply:
Yeah, this seems like an unnecessary indirection.